private/bufpkg/bufplugin/digest.go

```diff
@@ -0,0 +1,214 @@
+// Copyright 2020-2024 Buf Technologies, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package bufplugin
+
+import (
+	"bytes"
+	"encoding/hex"
+	"errors"
+	"fmt"
+	"strconv"
+	"strings"
+
+	"github.com/bufbuild/buf/private/bufpkg/bufcas"
+	"github.com/bufbuild/buf/private/bufpkg/bufparse"
+	"github.com/bufbuild/buf/private/pkg/syserror"
+)
+
+const (
+	// DigestTypeP1 represents the p1 plugin digest type.
+	//
+	// The string value of this is "p1".
+	DigestTypeP1 DigestType = iota + 1
+)
+
+var (
+	// AllDigestTypes are all known DigestTypes.
+	AllDigestTypes = []DigestType{
+		DigestTypeP1,
+	}
+	digestTypeToString = map[DigestType]string{
+		DigestTypeP1: "p1",
+	}
+	stringToDigestType = map[string]DigestType{
+		"p1": DigestTypeP1,
+	}
+)
+
+// DigestType is a type of digest.
+type DigestType int
+
+// ParseDigestType parses a DigestType from its string representation.
+//
+// This reverses DigestType.String().
+//
+// Returns an error of type *bufparse.ParseError if the string could not be parsed.
+func ParseDigestType(s string) (DigestType, error) {
```

Comment:
Had an offline conversation about whether `DigestType` should also be a shared package (e.g. `bufpkg/bufdigest`), since this is also duplicated across modules and plugins. The respective keys (e.g. `ModuleKey` and `PluginKey`) types would control for the digest type at the time of construction. Our conclusion is that this makes sense, but should it be done as a follow-up PR given the size of the migration of `Ref`/`FullName`?